File: web-go/internal/controller/user_controller.go
// internal/controller/user_controller.go
package controller

import (
	"net/http"

	"agussyahrilmubarok.github.io/web/internal/domain"
	"agussyahrilmubarok.github.io/web/internal/model"
	"agussyahrilmubarok.github.io/web/internal/service"
	"agussyahrilmubarok.github.io/web/pkg/helper"
	"github.com/gin-contrib/sessions"
	"github.com/gin-gonic/gin"
)

type userController struct {
	userService service.IUserService
}

func (h *userController) Index(c *gin.Context) {
	userProfile, ok := mustGetUserFromContext(c)
	if !ok {
		return
	}

	session := sessions.Default(c)
	data := gin.H{
		"Title":       "Users",
		"UserProfile": userProfile,
	}

	if flashes := session.Flashes("success"); len(flashes) > 0 {
		data["MsgInfo"] = flashes[0]
	}
	session.Save()

	users, err := h.userService.GetAll(c.Request.Context())
	if err != nil {
		data["MsgError"] = "Failed to load users"
		render(c, http.StatusInternalServerError, "users_index.html", data)
		return
	}

	data["Values"] = users
	render(c, http.StatusOK, "users_index.html", data)
}

func (h *userController) Create(c *gin.Context) {
	userProfile, ok := mustGetUserFromContext(c)
	if !ok {
		return
	}

	data := gin.H{
		"Title":       "Create User",
		"UserProfile": userProfile,
	}

	var req model.CreateUserRequest
	data["Values"] = req

	render(c, http.StatusOK, "users_create.html", data)
}

func (h *userController) Store(c *gin.Context) {
	userProfile, ok := mustGetUserFromContext(c)
	if !ok {
		return
	}

	data := gin.H{
		"Title":       "Create User",
		"UserProfile": userProfile,
	}

	var req model.CreateUserRequest
	if err := c.ShouldBind(&req); err != nil {
		data["Values"] = req
		data["Errors"] = helper.ValidatorError(err)
		render(c, http.StatusBadRequest, "users_create.html", data)
		return
	}

	_, err := h.userService.Create(c.Request.Context(), req)
	if err != nil {
		errorsMap := make(map[string]string)
		switch err {
		case domain.ErrUserEmailExists:
			errorsMap["Email"] = "Email already exists"
		default:
			errorsMap["Error"] = "Something went wrong"
		}
		data["Values"] = req
		data["Errors"] = errorsMap
		render(c, http.StatusBadRequest, "users_create.html", data)
		return
	}

	session := sessions.Default(c)
	session.AddFlash("User created successfully!", "success")
	session.Save()

	c.Redirect(http.StatusFound, "/dashboard/users")
}

// Edit — GET /dashboard/users/:id/edit
func (h *userController) Edit(c *gin.Context) {
	userProfile, ok := mustGetUserFromContext(c)
	if !ok {
		return
	}

	id := c.Param("id")
	user, err := h.userService.GetByID(c.Request.Context(), id)
	if err != nil || user == nil {
		session := sessions.Default(c)
		session.AddFlash("User not found", "error")
		session.Save()
		c.Redirect(http.StatusFound, "/dashboard/users")
		return
	}

	data := gin.H{
		"Title":       "Edit User",
		"UserProfile": userProfile,
		"Data":        user,
	}
	render(c, http.StatusOK, "users_edit.html", data)
}

// Update — POST /dashboard/users/:id/edit
func (h *userController) Update(c *gin.Context) {
	userProfile, ok := mustGetUserFromContext(c)
	if !ok {
		return
	}

	id := c.Param("id")

	var req model.UpdateUserRequest
	if err := c.ShouldBind(&req); err != nil {
		user, _ := h.userService.GetByID(c.Request.Context(), id)
		data := gin.H{
			"Title":       "Edit User",
			"UserProfile": userProfile,
			"Data":        user,
			"Errors":      helper.ValidatorError(err),
		}
		render(c, http.StatusBadRequest, "users_edit.html", data)
		return
	}

	_, err := h.userService.UpdateByID(c.Request.Context(), id, req)
	if err != nil {
		errorsMap := make(map[string]string)
		switch err {
		case domain.ErrUserNotFound:
			errorsMap["Error"] = "User not found"
		case domain.ErrUserEmailExists:
			errorsMap["Email"] = "Email already exists"
		default:
			errorsMap["Error"] = "Something went wrong"
		}
		user, _ := h.userService.GetByID(c.Request.Context(), id)
		data := gin.H{
			"Title":       "Edit User",
			"UserProfile": userProfile,
			"Data":        user,
			"Errors":      errorsMap,
		}
		render(c, http.StatusBadRequest, "users_edit.html", data)
		return
	}

	session := sessions.Default(c)
	session.AddFlash("User updated successfully!", "success")
	session.Save()

	c.Redirect(http.StatusFound, "/dashboard/users")
}

// Delete — POST /dashboard/users/:id/delete
func (h *userController) Delete(c *gin.Context) {
	id := c.Param("id")

	session := sessions.Default(c)

	if err := h.userService.DeleteByID(c.Request.Context(), id); err != nil {
		session.AddFlash("Failed to delete user", "error")
	} else {
		session.AddFlash("User deleted successfully!", "success")
	}
	session.Save()

	c.Redirect(http.StatusFound, "/dashboard/users")
}

func NewUserController(userService service.IUserService) *userController {
	return &userController{userService: userService}
}

File: examples/web-mvc-bootstrap/web-spring/src/main/java/io/github/agussyahrilmubarok/web/delivery/web/controller/UserController.java
package io.github.agussyahrilmubarok.web.delivery.web.controller;

import io.github.agussyahrilmubarok.web.application.dto.user.CreateUserRequest;
import io.github.agussyahrilmubarok.web.application.dto.user.UpdateUserRequest;
import io.github.agussyahrilmubarok.web.application.dto.user.UserResponse;
import io.github.agussyahrilmubarok.web.application.service.UserService;
import io.github.agussyahrilmubarok.web.common.exception.EmailAlreadyInUseException;
import io.github.agussyahrilmubarok.web.common.exception.NotFoundException;
import io.github.agussyahrilmubarok.web.common.WebUtils;
import io.github.agussyahrilmubarok.web.infrastructure.security.CustomUserDetails;
import jakarta.validation.Valid;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

@Controller
@RequestMapping("/dashboard/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    @GetMapping
    public String list(final Model model) {
        model.addAttribute("users", userService.getAll());
        return "dashboard/users/index";
    }

    @GetMapping("/create")
    public String createForm(@ModelAttribute("createUser") final CreateUserRequest createUserRequest) {
        return "dashboard/users/create";
    }

    @PostMapping("/create")
    public String createSubmit(
            @ModelAttribute("createUser") @Valid final CreateUserRequest createUserRequest,
            final BindingResult bindingResult,
            final Model model,
            final RedirectAttributes redirectAttributes) {
        if (bindingResult.hasErrors()) return "dashboard/users/create";

        try {
            userService.create(createUserRequest);
        } catch (final EmailAlreadyInUseException e) {
            bindingResult.rejectValue("email", "Exists.user.email");
            return "dashboard/users/create";
        } catch (final Exception e) {
            model.addAttribute(WebUtils.MSG_ERROR, "Something went wrong. Please try again");
            return "dashboard/users/create";
        }

        redirectAttributes.addFlashAttribute(WebUtils.MSG_SUCCESS, "User created successfully");
        return "redirect:/dashboard/users";
    }

    @GetMapping("/edit/{id}")
    public String editForm(
            @PathVariable final UUID id, final Model model, final RedirectAttributes redirectAttributes) {
        try {
            UserResponse user = userService.getById(id);
            UpdateUserRequest editUser = new UpdateUserRequest(user.id(), user.name(), user.email(), null);
            model.addAttribute("editUser", editUser);
        } catch (final NotFoundException e) {
            redirectAttributes.addFlashAttribute(WebUtils.MSG_ERROR, "User not found");
            return "redirect:/dashboard/users";
        }
        return "dashboard/users/edit";
    }

    @PostMapping("/edit/{id}")
    public String editSubmit(
            @PathVariable final UUID id,
            @ModelAttribute("editUser") @Valid final UpdateUserRequest request,
            final BindingResult bindingResult,
            final Model model,
            final RedirectAttributes redirectAttributes) {
        if (bindingResult.hasErrors()) return "dashboard/users/edit";

        try {
            userService.update(id, request);
        } catch (final EmailAlreadyInUseException e) {
            model.addAttribute("editUser", request);
            bindingResult.rejectValue("email", "Exists.user.email", "Email is already in use");
            return "dashboard/users/edit";
        } catch (final NotFoundException e) {
            redirectAttributes.addFlashAttribute(WebUtils.MSG_ERROR, "User not found");
            return "redirect:/dashboard/users";
        } catch (final Exception e) {
            model.addAttribute("editUser", request);
            model.addAttribute(WebUtils.MSG_ERROR, "Something went wrong. Please try again");
            return "dashboard/users/edit";
        }

        redirectAttributes.addFlashAttribute(WebUtils.MSG_SUCCESS, "User updated successfully");
        return "redirect:/dashboard/users";
    }

    @PostMapping("/delete/{id}")
    public String delete(@PathVariable final UUID id, final RedirectAttributes redirectAttributes) {
        try {
            userService.delete(id);
            redirectAttributes.addFlashAttribute(WebUtils.MSG_SUCCESS, "User deleted successfully");
        } catch (final NotFoundException e) {
            redirectAttributes.addFlashAttribute(WebUtils.MSG_ERROR, "User not found");
        }
        return "redirect:/dashboard/users";
    }

    @ModelAttribute
    public void globalAttributes(@AuthenticationPrincipal final CustomUserDetails userDetails, final Model model) {
        if (userDetails != null) {
            model.addAttribute("userProfile", userDetails.getUser());
        }
    }
}
